proxy/src/cancellation.rs

```diff
@@ -79,13 +95,33 @@ impl<P: CancellationPublisher> CancellationHandler<P> {
             cancellation_handler: self,
         }
     }
+
     /// Try to cancel a running query for the corresponding connection.
     /// If the cancellation key is not found, it will be published to Redis.
+    /// check_allowed - if true, check if the IP is allowed to cancel the query
     pub(crate) async fn cancel_session(
         &self,
         key: CancelKeyData,
         session_id: Uuid,
+        peer_addr: &IpAddr,
+        check_allowed: bool,
     ) -> Result<(), CancelError> {
+        // TODO: check for unspecified address is only for backward compatibility, should be removed
+        if !peer_addr.is_unspecified() {
+            let subnet_key = ipnet::IpNet::from(*peer_addr).netmask();
```

Comment:
This doesn't look correct to me. `netmask` would likely always return `255.255.255.255` here.

I think this is correct, with some basic tests run locally, matching what you had before

```rust
match *peer_addr {
    IpAddr::V4(ip) => IpNet::V4(Ipv4Net::new_assert(ip, 24).trunc()),
    IpAddr::V6(ip) => IpNet::V6(Ipv6Net::new_assert(ip, 64).trunc()),
}
```
(need to update ipnet to get `new_assert`).

I also wonder if `/24` is too restrictive. Perhaps just `/32` for ipv4 is good enough.

Reply:
Thanks! This could have been dangerous. I didn't check the internals of the netmask() cause thought that it would have something different when we don't have any mask. I'll update